CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+--Panel3D: 3D GUI panels for VR
+--written by 0xBAADF00D
+--revised/refactored 5/11/16
+
+local UserInputService = game:GetService("UserInputService")
+local RunService = game:GetService("RunService")
+local GuiService = game:GetService("GuiService")
+local CoreGui = game:GetService("CoreGui")
+local RobloxGui = CoreGui:WaitForChild("RobloxGui")
+
+
+--Util
+local Util = {}
+do
+	function Util.Create(instanceType)
+		return function(data)
+			local obj = Instance.new(instanceType)
+			for k, v in pairs(data) do
+				if type(k) == 'number' then
+					v.Parent = obj
+				else
+					obj[k] = v
+				end
+			end
+			return obj
+		end
+	end
+
+	-- RayPlaneIntersection (shortened)
+	-- http://www.siggraph.org/education/materials/HyperGraph/raytrace/rayplane_intersection.htm
+	function Util.RayPlaneIntersection(ray, planeNormal, pointOnPlane)
+		planeNormal = planeNormal.unit
+		ray = ray.Unit
+
+		local Vd = planeNormal:Dot(ray.Direction)
+		if Vd == 0 then -- parallel, no intersection
+			return nil
+		end
+
+		local V0 = planeNormal:Dot(pointOnPlane - ray.Origin)
+		local t = V0 / Vd
+		if t < 0 then --plane is behind ray origin, and thus there is no intersection
+			return nil
+		end
+		
+		return ray.Origin + ray.Direction * t
+	end
+end
+--End of Util
+
+
+--Panel3D State variables
+local coreScriptMode = true
+local renderStepName = "Panel3DRenderStep-" .. game:GetService("HttpService"):GenerateGUID()
+local defaultPixelsPerStud = 64
+local pointUpCF = CFrame.Angles(math.rad(-90), math.rad(180), 0)
+local zeroVector = Vector3.new(0, 0, 0)
+local zeroVector2 = Vector2.new(0, 0)
+local turnAroundCF = CFrame.Angles(0, math.rad(180), 0)
+local fullyOpaqueAtPixelsFromEdge = 10
+local fullyTransparentAtPixelsFromEdge = 80
+
+local cursorHidden = false
+local cursorHideTime = 2.5
+
+local currentModal = nil
+local lastModal = nil
+local currentMaxDist = math.huge
+local currentClosest = nil
+local currentHeadScale = 1
+local panels = {}
+local headHeightFromFloor = 5
```

Comment:
Is this something that should be hardcoded at 5? What if the game we are playing has a larger character?